### PR TITLE
[Merged by Bors] - Spark anomaly detection demo

### LINKED ISSUE #142: Anomaly detection demo: Spark-k8s
https://github.com/stackabletech/stackablectl/issues/142#tasklist-block-775312d2-f3c0-4bf1-a5e4-6ad7a6b61e0c

### PR DESCRIPTION
**Pull request by adwk67**
## Description

A demo using spark-k8s to create anomaly detection scores from NYC taxi data, writing the results to a Trino table and displaying them in a Superset dashboard.

Closes #142. 

## Review Checklist
- [ ] Code contains useful comments
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by adwk67** [comment]
Sample output: 
image

File: demos/demos-v1.yaml [hunk 1]
```diff
@@ -53,6 +53,21 @@ demos:
       - plainYaml: https://raw.githubusercontent.com/stackabletech/stackablectl/main/demos/nifi-kafka-druid-water-level-data/create-nifi-ingestion-job.yaml
       - plainYaml: https://raw.githubusercontent.com/stackabletech/stackablectl/main/demos/nifi-kafka-druid-water-level-data/create-druid-ingestion-job.yaml
       - plainYaml: https://raw.githubusercontent.com/stackabletech/stackablectl/main/demos/nifi-kafka-druid-water-level-data/setup-superset.yaml
+  spark-k8s-anomaly-detection-taxi-data:
+    description: Demo loading New York taxi data into an S3 bucket and carrying out an anomaly detection anlysis on it
```
Comment (sbernauer) [inline demos/demos-v1.yaml:57]:
```suggestion
    description: Demo loading New York taxi data into an S3 bucket and carrying out an anomaly detection analysis on it
```

File: demos/spark-k8s-anomaly-detection-taxi-data/create-spark-anomaly-detection-job.yaml [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+data:
+  spark-ad-job.yaml: |
+    ---
+    apiVersion: spark.stackable.tech/v1alpha1
+    kind: SparkApplication
+    metadata:
+      name: spark-ad
+    spec:
+      version: "1.0"
+      sparkImage: docker.stackable.tech/aken/pyspark-k8s-with-iceberg:0.1.0 #docker.stackable.tech/stackable/pyspark-k8s:3.3.0-stackable0.2.0
```
Comment (sbernauer) [inline demos/spark-k8s-anomaly-detection-taxi-data/create-spark-anomaly-detection-job.yaml:40]:
```suggestion
      sparkImage: docker.stackable.tech/demos/pyspark-k8s-with-kafka-and-iceberg:3.3.0-stackable0.2.0 # docker.stackable.tech/stackable/pyspark-k8s:3.3.0-stackable0.2.0
```

File: demos/spark-k8s-anomaly-detection-taxi-data/create-spark-anomaly-detection-job.yaml [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+      driver:
+        volumeMounts:
+          - name: cm-spark
+            mountPath: /spark-scripts
+      executor:
+        volumeMounts:
+          - name: cm-spark
+            mountPath: /spark-scripts
+      sparkConf:
+        spark.kubernetes.submission.waitAppCompletion: "false"
```
Comment (sbernauer) [inline demos/spark-k8s-anomaly-detection-taxi-data/create-spark-anomaly-detection-job.yaml:61]: What do you think about using the new structs added in https://github.com/stackabletech/spark-k8s-operator/pull/166?
Additionally, please consider giving the submit-job more resources than below. It was toking so long, i started to look into whats the problem ^^
```
    Limits:                                                                                                                                                                                                    
      cpu:     100m                                                                                                                                                                                            
      memory:  1Gi                                                                                                                                                                                             
    Requests:                                                                                                                                                                                                  
      cpu:     50m                                                                                                                                                                                             
      memory:  1Gi
```

Comment (adwk67) [reply demos/spark-k8s-anomaly-detection-taxi-data/create-spark-anomaly-detection-job.yaml:78]: Done!

**Review by sbernauer — APPROVED** [review]
Demo worked fine for me :+1: 
Iceberg might be a bit of overkill here, but maybe gives us benefits in the future, e.g. compaction during streaming.

File: docs/modules/ROOT/pages/demos/spark-k8s-anomaly-detection-taxi-data.adoc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+You can see a folder (called prefixes in S3) containing a dataset of similarly-structured data files. The data is partitioned by month and contains several hundred MBs of data. This may not seem particularly large for a data-set, but the model is a time-series model where the data has decreasing relevance the "older" it is: this is especially when the data is subject to multiple external factors, many of which are unknown and fluctuating in scope and effect.
+
+
+The second bucket `prediction` contains the output from the model scoring process:
+
+image::spark-k8s-anomaly-detection-taxi-data/minio_4.png[]
+
+This is a much smaller file as it only contains scores for each aggregated time period.
+
+== Spark
```
Comment (sbernauer) [inline docs/modules/ROOT/pages/demos/spark-k8s-anomaly-detection-taxi-data.adoc:90]: Mentioning `kubectl port-forward spark-ad-driver 4040` and a screenshot would be great

File: demos/spark-k8s-anomaly-detection-taxi-data/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,10 @@
+FROM docker.stackable.tech/stackable/pyspark-k8s:3.3.0-stackable0.2.0
```
[the remark this answers is not in the document]
Comment (adwk67) [reply demos/spark-k8s-anomaly-detection-taxi-data/Dockerfile]: Removed.

File: docs/modules/ROOT/pages/demos/spark-k8s-anomaly-detection-taxi-data.adoc (hunk 4 above)
Comment (adwk67) [reply docs/modules/ROOT/pages/demos/spark-k8s-anomaly-detection-taxi-data.adoc:89]: Added.

**Review by sbernauer — APPROVED** [review]

**Comment by adwk67** [comment]
bors merge

**Pull request closed without merging**